File: src/css/rules/unknown.zig
pub const css = @import("../css_parser.zig");
pub const css_values = @import("../values/values.zig");
pub const Error = css.Error;
const Printer = css.Printer;
const PrintErr = css.PrintErr;

/// An unknown at-rule, stored as raw tokens.
pub const UnknownAtRule = struct {
    /// The name of the at-rule (without the @).
    name: []const u8,
    /// The prelude of the rule.
    prelude: css.TokenList,
    /// The contents of the block, if any.
    block: ?css.TokenList,
    /// The location of the rule in the source file.
    loc: css.Location,

    const This = @This();

    pub fn toCss(this: *const This, dest: *Printer) PrintErr!void {
        // #[cfg(feature = "sourcemap")]
        // dest.add_mapping(self.loc);

        try dest.writeChar('@');
        try dest.writeStr(this.name);

        if (this.prelude.v.items.len > 0) {
            try dest.writeChar(' ');
            try this.prelude.toCss(dest, false);
        }

        if (this.block) |*block| {
            try dest.whitespace();
            try dest.writeChar('{');
            dest.indent();
            try dest.newline();
            try block.toCss(dest, false);
            dest.dedent();
            try dest.newline();
            try dest.writeChar('}');
        } else {
            try dest.writeChar(';');
        }
    }

    pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) This {
        return css.implementDeepClone(@This(), this, allocator);
    }
};

const std = @import("std");
const Allocator = std.mem.Allocator;

File: src/css/rules/unknown.rs
use crate::css_rules::Location;
use crate::properties::custom::TokenList;
use crate::{PrintErr, Printer};

/// An unknown at-rule, stored as raw tokens.
pub struct UnknownAtRule {
    /// The name of the at-rule (without the @).
    // TODO: arena lifetime — slice backed by parser arena.
    pub(crate) name: &'static [u8],
    /// The prelude of the rule.
    pub(crate) prelude: TokenList,
    /// The contents of the block, if any.
    pub(crate) block: Option<TokenList>,
    /// The location of the rule in the source file.
    pub(crate) loc: Location,
}

impl UnknownAtRule {
    pub(crate) fn to_css(&self, dest: &mut Printer) -> Result<(), PrintErr> {
        // #[cfg(feature = "sourcemap")]
        // dest.add_mapping(self.loc);

        dest.write_char(b'@')?;
        dest.write_str(self.name)?;

        if !self.prelude.v.is_empty() {
            dest.write_char(b' ')?;
            self.prelude.to_css(dest)?;
        }

        if let Some(block) = &self.block {
            dest.block(|d| {
                d.newline()?;
                block.to_css(d)
            })
        } else {
            dest.write_char(b';')
        }
    }

    pub(crate) fn deep_clone(&self, bump: &bun_alloc::Arena) -> Self {
        use crate::generics::DeepClone as _;
        // `name` is an arena-owned slice → identity copy; `TokenList`
        // carries `#[derive(DeepClone)]`.
        Self {
            name: self.name,
            prelude: self.prelude.deep_clone(bump),
            block: self.block.as_ref().map(|b| b.deep_clone(bump)),
            loc: self.loc,
        }
    }
}
